### PR TITLE
snap: introduce a struct Channel to represent store channels, and helpers to work with it

### PR DESCRIPTION
**Pull request by pedronis**
This introduces snap.Channel to represent store channels, with ParseChannel for parsing it, and helpers to turn it back into strings and to check whether two channels match.

### CODE REVIEW

File: snap/channel.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	switch len(p) {
+	default:
+		return StoreChannel{}, fmt.Errorf("channel name has too many components: %s", s)
+	case 3:
+		track, risk, branch = p[0], p[1], p[2]
+	case 2:
+		if strutil.ListContains(channelRisks, p[0]) {
+			risk, branch = p[0], p[1]
+		} else {
+			track, risk = p[0], p[1]
```
Comment (bboozzoo) [inline snap/channel.go]: Is is possible to do some validation of track/branch or are these basically freeform?

Comment (pedronis) [reply snap/channel.go]: we could but essentially mostly the store cares, we still be able to produce reasonable errors, actually this is already tying us to things for which mostly only  the store should care at it is

File: snap/channel_test.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+func (s *storeChannelSuite) TestFull(c *C) {
+	tests := []struct {
+		channel string
+		str     string
+	}{
+		{"stable", "latest/stable"},
+		{"latest/stable", "latest/stable"},
+		{"1.0/edge", "1.0/edge"},
```
Comment (chipaca) [inline snap/channel_test.go]: I think you're missing the `"foo"` -> `"foo/stable"` case

File: snap/channel.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/snapcore/snapd/arch"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var channelRisks = []string{"stable", "candidate", "beta", "edge"}
+
+// Channel identifies and describes completely a store channel.
```
Comment (mvo5) [inline snap/channel.go]: (nitpick) To my ears `Channel icompletely dentifies and describes a store channel.` sounds a tad more natural, i.e. changing the position of "completely".
`

File: snap/channel.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+// ParseChannel parses a string representing a store channel and includes the given architecture, if architecture is "" the system architecture is included.
+func ParseChannel(s string, architecture string) (Channel, error) {
+	if s == "" {
+		return Channel{}, fmt.Errorf("channel name cannot be empty")
+	}
+	p := strings.Split(s, "/")
+	var risk, track, branch string
+	switch len(p) {
```
Comment (mvo5) [inline snap/channel.go]: Nice way to parse this!

File: snap/channel.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+type Channel struct {
+	Architecture string `json:"architecture"`
+	Name         string `json:"name"`
+	Track        string `json:"track"`
+	Risk         string `json:"risk"`
+	Branch       string `json:"branch,omitempty"`
+}
+
+// ParseChannel parses a string representing a store channel and includes the given architecture, if architecture is "" the system architecture is included.
+func ParseChannel(s string, architecture string) (Channel, error) {
```
Comment (mvo5) [inline snap/channel.go]: What will `architecture` be used for? I.e. when do we need this?

File: snap/channel.go (hunk 3 above)
Comment (pedronis) [reply snap/channel.go]: @chipaca opinion on this? I'm tempted to merge and adjust in the/a follow up in case

Comment (chipaca) [reply snap/channel.go]: `Channel identifies and completely describes` -> does not completely identify
`Channel completely identifies and describes` -> completely does both
`Channel completely identifies, and describes` -> probably both, maybe just 1st
`Channel identifies and describes completely` -> probably just the second, maybe both
`Channel identifies and describes, completely` -> both again

take your pick

**Pull request merged**